File: engine/src/flutter/lib/web_ui/lib/src/engine/compositor/color_filter.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// @dart = 2.10
part of engine;

/// A [ui.ColorFilter] backed by Skia's [CkColorFilter].
class CkColorFilter extends ManagedSkiaObject<SkColorFilter> {
  final EngineColorFilter _engineFilter;

  CkColorFilter.mode(EngineColorFilter filter) : _engineFilter = filter;

  CkColorFilter.matrix(EngineColorFilter filter) : _engineFilter = filter;

  CkColorFilter.linearToSrgbGamma(EngineColorFilter filter)
      : _engineFilter = filter;

  CkColorFilter.srgbToLinearGamma(EngineColorFilter filter)
      : _engineFilter = filter;

  SkColorFilter _createSkiaObjectFromFilter() {
    SkColorFilter skColorFilter;
    switch (_engineFilter._type) {
      case EngineColorFilter._TypeMode:
        skColorFilter = canvasKit.SkColorFilter.MakeBlend(
          toSharedSkColor1(_engineFilter._color!),
          toSkBlendMode(_engineFilter._blendMode!),
        );
        break;
      case EngineColorFilter._TypeMatrix:
        final Float32List colorMatrix = Float32List(20);
        final List<double> matrix = _engineFilter._matrix!;
        for (int i = 0; i < 20; i++) {
          colorMatrix[i] = matrix[i];
        }
        skColorFilter = canvasKit.SkColorFilter.MakeMatrix(colorMatrix);
        break;
      case EngineColorFilter._TypeLinearToSrgbGamma:
        skColorFilter = canvasKit.SkColorFilter.MakeLinearToSRGBGamma();
        break;
      case EngineColorFilter._TypeSrgbToLinearGamma:
        skColorFilter = canvasKit.SkColorFilter.MakeSRGBToLinearGamma();
        break;
      default:
        throw StateError(
            'Unknown mode ${_engineFilter._type} for ColorFilter.');
    }
    return skColorFilter;
  }

  @override
  SkColorFilter createDefault() {
    return _createSkiaObjectFromFilter();
  }

  @override
  SkColorFilter resurrect() {
    return _createSkiaObjectFromFilter();
  }

  @override
  void delete() {
    rawSkiaObject?.delete();
  }
}

File: lib/ui/painting/color_filter.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/ui/painting/color_filter.h"

#include <cstring>

#include "flutter/lib/ui/ui_dart_state.h"
#include "third_party/tonic/converter/dart_converter.h"
#include "third_party/tonic/dart_args.h"
#include "third_party/tonic/dart_binding_macros.h"
#include "third_party/tonic/dart_library_natives.h"

namespace flutter {

IMPLEMENT_WRAPPERTYPEINFO(ui, ColorFilter);

void ColorFilter::Create(Dart_Handle wrapper) {
  UIDartState::ThrowIfUIOperationsProhibited();
  auto res = fml::MakeRefCounted<ColorFilter>();
  res->AssociateWithDartWrapper(wrapper);
}

void ColorFilter::initMode(int color, int blend_mode) {
  filter_ = std::make_shared<DlBlendColorFilter>(
      static_cast<DlColor>(color), static_cast<DlBlendMode>(blend_mode));
}

void ColorFilter::initMatrix(const tonic::Float32List& color_matrix) {
  FML_CHECK(color_matrix.num_elements() == 20);

  // Flutter still defines the matrix to be biased by 255 in the last column
  // (translate). skia is normalized, treating the last column as 0...1, so we
  // post-scale here before calling the skia factory.
  float matrix[20];
  memcpy(matrix, color_matrix.data(), sizeof(matrix));
  matrix[4] *= 1.0f / 255;
  matrix[9] *= 1.0f / 255;
  matrix[14] *= 1.0f / 255;
  matrix[19] *= 1.0f / 255;
  filter_ = std::make_shared<DlMatrixColorFilter>(matrix);
}

void ColorFilter::initLinearToSrgbGamma() {
  filter_ = DlLinearToSrgbGammaColorFilter::instance;
}

void ColorFilter::initSrgbToLinearGamma() {
  filter_ = DlSrgbToLinearGammaColorFilter::instance;
}

ColorFilter::~ColorFilter() = default;

}  // namespace flutter
